### PR TITLE
test(ci): signoff csv gates v0

### PR DESCRIPTION
**Pull request by plitenh**
## What Changed

- Extract `chipcompiler.engine.signoff.csv` for profile-driven CSV export, gate evaluation, FileCheck projection, and
`run_manifest.json`.
- Add lit suite under `test/lit/` (ready + negative fixtures; optional packaged workspace) and thin `.github/scripts`
entrypoints.
- Wire fixture lit into default PR CI; document v0 / trial status and graduation bar in `docs/signoff-csv-gates.md`.
- Nix pins for lit/filecheck + `nix run .#ci-signoff-lit` (checkout + `.venv` still required for chipcompiler).

## Scope

Select the areas touched by this PR:

- [ ] CLI - command behavior, Typer command surface, output formats, or workspace commands.
- [ ] Flow/runtime - workspace lifecycle, EngineFlow, step execution, logs, metrics, or artifacts.
- [ ] EDA integration - Yosys, ECC-Tools, DreamPlace, KLayout, PDKs, or native/runtime wrappers.
- [x] Build/package - Nix, PyInstaller, wheels, `uv.lock`, or release artifacts.
- [x] CI/release - GitHub Actions, version checks, changelog, or release automation.
- [x] Tests/docs only

## Runtime And Packaging Impact

- [ ] No runtime or packaging impact
- [ ] CLI output or machine-readable contract changed
- [ ] Workspace layout, flow state, or artifact paths changed
- [ ] Native toolchain or wrapper behavior changed
- [ ] `ecc-tools` or `ecc-dreamplace` dependency changed
- [x] PyInstaller, Nix, or release artifact changed

Notes:

- Adds library under `chipcompiler.engine.signoff.csv` (not a new `ecc` CLI command).
- Nix flake exposes lit/filecheck / `ci-signoff-lit` / `ci-run-ics55-gcd`; default PR CI runs **fixture** lit only (no real
ics55 rtl2gds).
- CI contract profile: `test/lit/profiles/signoff.yml` (v0). Product readiness remains `ecc signoff inspect/export`.

## Validation

List the commands you ran. Mark checks that are not applicable as N/A.

- [x] `uv run pytest test/`
- [x] `uv run ruff check chipcompiler test`
- [x] `uv run ruff format --check chipcompiler test`
- [ ] PyInstaller smoke: `ecc --help`, `ecc --version`, `ecc version --json` — N/A (no bundle/CLI surface change)
- [ ] Nix smoke: `nix run .#cli -- --help` — N/A
- [x] Manual flow smoke:
  - `uv run pytest test/test_signoff_csv_export.py -q`
  - `ECC_REPO_ROOT=$PWD uv run --with lit==18.1.8 --with filecheck==1.0.6 python .github/scripts/ci_signoff_lit.py`
- [x] Other: default CI step `Signoff CSV lit gates` (fixture-only)

Skipped checks and reason:

- Packaged / real-PDK path (`ci-run-ics55-gcd` + `ECC_TEST_WORKSPACE`) intentionally optional; not required for this v0 PR.

## Checklist

- [x] I kept the change scoped to ECC.
- [x] I updated docs or user-facing CLI text where behavior changed.
- [ ] I included lockfile or version metadata updates when dependencies changed. — N/A (lit/filecheck via `uv run --with` / Nix
pins, not lockfile deps)
- [ ] I documented any submodule updates and why they are needed. — N/A
- [x] I did not include local caches, virtual environments, or generated build outputs.
- [x] I explained skipped validation and remaining risk.

### Remaining risk / v0 note

This is a **trial / v0 CI contract**, not Harden package truth. Formal check-layer graduation criteria are in
`docs/signoff-csv-gates.md` (owned reds, schema freeze, consumer). Follow-ups (nightly black-box, dashboard baseline) stay out
of this PR.

### CODE REVIEW

**Comment by plitenh** [comment]
CI passed on https://github.com/plitenh/ecc/actions/runs/35502889770

**Comment by plitenh** [comment]
@Emin017

File: .github/scripts/ci_signoff_lit.py
Comment (Emin017) [inline .github/scripts/ci_signoff_lit.py:1]: This script can be rewritten as a shell script.

File: .github/scripts/ci_run_ics55_gcd.sh
Comment (Emin017) [inline .github/scripts/ci_run_ics55_gcd.sh:1]: Consider supporting the ability to specify an existing `ecc.toml` and pass in the RTL design path, which would make it easier to support batch-running flows later.

File: .github/scripts/materialize_signoff_workspace.py
Comment (Emin017) [inline .github/scripts/materialize_signoff_workspace.py:1]: Why not use `jq`?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -125,6 +125,13 @@ jobs:
       - name: Pytest
         run: uv run --no-sync pytest test/ -v --basetemp=pytest-tmp --cov=chipcompiler --cov-report=
 
+      - name: Signoff CSV lit gates
+        run: |
+          # Fixture-only lit suite (no ics55 rtl2gds). lit/filecheck pinned to nix pins.
+          ECC_REPO_ROOT="$PWD" \
+            uv run --no-sync --with lit==18.1.8 --with filecheck==1.0.6 \
```
Comment (Emin017) [inline .github/workflows/ci.yml:134]: Pin lit and FileCheck by using `llvmPackages_23.libllvm` in [nixpkgs](https://search.nixos.org/packages?channel=26.05&query=llvmPackages_23.libllvm#show=llvmPackages_23.libllvm).

File: chipcompiler/engine/signoff/csv/compare.py
Comment (Emin017) [inline chipcompiler/engine/signoff/csv/compare.py:1]: Consider not adding this feature to ecc. Could it be implemented externally using [csvkit](https://csvkit.readthedocs.io/en/latest/)?

File: docs/signoff-csv-gates.md
Comment (Emin017) [inline docs/signoff-csv-gates.md:1]: Move this documentation to `test/lit/README.md`.

**Review by Emin017 — CHANGES_REQUESTED** [review]

File: chipcompiler/engine/signoff/csv/compare.py
Comment (plitenh) [reply chipcompiler/engine/signoff/csv/compare.py:1]: Already remove the "run-to-run comparison / dashboard diff" API from ecc, and keep only the minimal logic required for the CI gate.

File: .github/scripts/materialize_signoff_workspace.py
Comment (plitenh) [reply .github/scripts/materialize_signoff_workspace.py:1]: It is recommended to adopt this; it has already been changed to jq.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Emin017) [reply .github/workflows/ci.yml]: Why do we still need to install lit and FileCheck via Python here?

**Review by Emin017 — CHANGES_REQUESTED** [review]

**Comment by plitenh** [comment]
Addressed the Nix lit/FileCheck review notes:

- CI no longer installs lit/filecheck via `uv --with` (PyPI).
- manylinux Test job now runs `ci_install_nix.sh` (Nix is not in the image), then `ci_signoff_lit_nix.sh`.
- That enters a pinned nixpkgs shell with `lit` + `llvmPackages_23.llvm` (provides `FileCheck`; `llvm-lit` is not on that bin output in current nixpkgs) + `jq`.
- `nix/signoff-tools.nix` wraps the same FileCheck source for local `nix run .#ci-signoff-lit`.

Pinned nixpkgs for the CI lit shell: `44a91898084f46797b5fac650c7e8c9ac38c43d4` (llvm-23.1.0). Override with `NIXPKGS_REF` if needed.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (plitenh) [reply .github/workflows/ci.yml]: This is an oversight. Tomorrow I’ll move the lit step into Nix CI and use Nix-provided lit + FileCheck, without PyPI.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -125,6 +125,16 @@ jobs:
       - name: Pytest
         run: uv run --no-sync pytest test/ -v --basetemp=pytest-tmp --cov=chipcompiler --cov-report=
 
+      - name: Install Nix
+        run: bash .github/scripts/ci_install_nix.sh
```
Comment (Emin017) [inline .github/workflows/ci.yml]:
```suggestion
       uses: cachix/install-nix-action@v31
```

File: nix/signoff.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Flake surface: pins/apps only; ops in .github/scripts, tests in test/lit.
+# lit/FileCheck come from llvmPackages_23 (see signoff-tools.nix).
+
+{ pkgs }:
+
+let
+  llvmPackages_23 =
+    if pkgs ? llvmPackages_23 then pkgs.llvmPackages_23
```
Comment (Emin017) [inline nix/signoff.nix]: Don't be overly defensive.

File: nix/signoff-tools.nix [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+#
+# CI (manylinux) installs Nix then nix-shells these two + jq —
+# see .github/scripts/ci_install_nix.sh and ci_signoff_lit_nix.sh.
+# Do not install the PyPI lit/filecheck packages in CI.
+
+{
+  lib,
+  lit,
+  llvmPackages_23 ? null,
+  llvmPackages ? null,
```
Comment (Emin017) [inline nix/signoff-tools.nix]: Don't be overly defensive.

File: .github/scripts/export_signoff_csv.sh [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+root="${ECC_REPO_ROOT:-$PWD}"
+here="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+if [[ -n "${PYTHON:-}" ]]; then py="$PYTHON"
+elif [[ -x "$root/.venv/bin/python" ]]; then py="$root/.venv/bin/python"
+else py="python3"; fi
+
+export PYTHONPATH="$root${PYTHONPATH:+:$PYTHONPATH}"
+export ECC_REPO_ROOT="$root"
```
Comment (Emin017) [inline .github/scripts/export_signoff_csv.sh]: Download the PyInstaller packaged artifacts from the previous CI build artifacts instead of rebuilding them.

File: nix/signoff-tools.nix [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    runtimeInputs = [ llvm ];
+    text = ''
+      exec FileCheck "$@"
+    '';
+  };
+
+  exportCsvSh = ../.github/scripts/export_signoff_csv.sh;
+  materializeSh = ../.github/scripts/materialize_signoff_workspace.sh;
+  ciRunSh = ../.github/scripts/ci_run_ics55_gcd.sh;
+  ciSignoffLitSh = ../.github/scripts/ci_signoff_lit.sh;
```
Comment (Emin017) [inline nix/signoff-tools.nix]: Move these scripts to `nix/`.

**Review by Emin017 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -125,6 +125,16 @@ jobs:
       - name: Pytest
         run: uv run --no-sync pytest test/ -v --basetemp=pytest-tmp --cov=chipcompiler --cov-report=
 
+      - name: Install Nix
+        run: curl -fsSL https://install.determinate.systems/nix | sh -s -- install linux --init none --no-confirm
```
Comment (Emin017) [inline .github/workflows/ci.yml]:
```suggestion
        uses: cachix/install-nix-action@v31
    - name: Magic Nix cache
      uses: DeterminateSystems/magic-nix-cache-action@v15
```

File: nix/ci_run_ics55_gcd.sh
Comment (Emin017) [inline nix/ci_run_ics55_gcd.sh]: Rename this file to `run_design.sh`

**Review by Emin017 — CHANGES_REQUESTED** [review]
It looks mostly good. What’s still needed is: move all the shell scripts and Python scripts under `nix/` into `nix/scripts` and remove the `ci_` prefix, then remove the pytest tests that are testing the csv scripts themselves (these should have been deleted after the test functionality was verified).

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -125,6 +125,16 @@ jobs:
...
+
+      - name: Signoff CSV lit gates
```
Comment (Emin017) [inline .github/workflows/ci.yml]: Run the signoff gates check after the PyInstaller CI build, so it can execute in parallel with pytest and save time.

File: flake.nix [hunk 9]
```diff
@@ -153,14 +153,22 @@
...
-      packages.default = pkgs.callPackage chipcompiler {
-        ecc-dreamplace = ecc-dreamplace.packages.${system}.default;
-        ecc-tools = ecc-tools.packages.${system}.default;
-        jsonrpcserver = pkgs.callPackage jsonrpcserver { oslash = pkgs.callPackage oslash {}; };
-        rosettakit = pkgs.callPackage rosettakit {};
-        yosysWithSlang = infra.packages.${system}.yosysWithSlang;
-      };
+    perSystem = { self', pkgs, system, ... }:
+      let
+        signoff = import ./nix/signoff.nix { inherit pkgs; };
```
Comment (Emin017) [inline flake.nix]:
```suggestion
        imports = [ ./nix/signoff.nix ];
```
See: https://flake.parts/options/flake-parts-modules.html

**Review by Emin017 — CHANGES_REQUESTED** [review]

**Comment by plitenh** [comment]
> It looks mostly good. What’s still needed is: move all the shell scripts and Python scripts under `nix/` into `nix/scripts` and remove the `ci_` prefix, then remove the pytest tests that are testing the csv scripts themselves (these should have been deleted after the test functionality was verified).

 Thanks — will do. This cleanup is a prerequisite for wiring lit into the signoff path.

File: .github/workflows/ci.yml (hunk 8 above)
Comment (plitenh) [reply .github/workflows/ci.yml]: Thanks for the suggestion. We’ll optimize CI runtime and parallelization after lit is stable and signoff is done.

**Review by Emin017 — APPROVED** [review]
